packages/flutter/lib/src/material/scrollbar.dart

```diff
@@ -0,0 +1,168 @@
+
+  void paintScrollbar(Canvas canvas, Size size) {
+    Point thumbOrigin;
+    Size thumbSize;
+
+    switch (scrollDirection) {
+      case Axis.vertical:
+        double thumbHeight = size.height * containerExtent / contentExtent;
+        thumbHeight = thumbHeight.clamp(_kMinScrollbarThumbLength, size.height);
+        final double maxThumbTop = size.height - thumbHeight;
+        double thumbTop = (scrollOffset / (contentExtent - containerExtent)) * maxThumbTop;
+        thumbTop = thumbTop.clamp(0.0, maxThumbTop);
+        thumbOrigin = new Point(size.width - _kScrollbarThumbGirth, thumbTop);
+        thumbSize = new Size(_kScrollbarThumbGirth, thumbHeight);
+        break;
+      case Axis.horizontal:
+        double thumbWidth = size.width * containerExtent / contentExtent;
+        thumbWidth = thumbWidth.clamp(_kMinScrollbarThumbLength, size.width);
+        final double maxThumbLeft = size.width - thumbWidth;
+        double thumbLeft = (scrollOffset / (contentExtent - containerExtent)) * maxThumbLeft;
+        thumbLeft = thumbLeft.clamp(0.0, maxThumbLeft);
+        thumbOrigin = new Point(thumbLeft, size.height - _kScrollbarThumbGirth);
+        thumbSize = new Size(thumbWidth, _kScrollbarThumbGirth);
+        break;
+    }
+
+    final Paint paint = new Paint()..color = thumbColor;
+    canvas.drawRect(thumbOrigin & thumbSize, paint);
+  }
+
+  @override
+  void paint(Canvas canvas, Size size) {
+    if (scrollOffset == null || thumbColor.alpha == 0)
+      return;
+    paintScrollbar(canvas, size);
+  }
+
+  @override
+  bool shouldRepaint(_ScrollbarPainter oldPainter) {
+    return oldPainter.scrollOffset != scrollOffset
+      || oldPainter.scrollDirection != scrollDirection
+      || oldPainter.contentExtent != contentExtent
+      || oldPainter.containerExtent != containerExtent
+      || oldPainter.thumbColor != thumbColor;
+  }
+}
+
+/// Displays a scrollbar that tracks the scrollOffset of its child's [Scrollable]
+/// descendant. If the Scrollbar's child has more than one Scrollable descendant
+/// the scrollableKey parameter can be used to identify the one the Scrollbar
+/// should track.
+class Scrollbar extends StatefulWidget {
+  Scrollbar({ Key key, this.scrollableKey, this.child }) : super(key: key) {
+    assert(child != null);
+  }
+
+  final Key scrollableKey;
+  final Widget child;
+
+  @override
+  _ScrollbarState createState() => new _ScrollbarState();
+}
+
+class _ScrollbarState extends State<Scrollbar> {
+  AnimationController _fade = new AnimationController(duration: _kScrollbarThumbFadeDuration);
+  CurvedAnimation _opacity;
+  double _scrollOffset;
+  Axis _scrollDirection;
+  double _containerExtent;
+  double _contentExtent;
+
+  @override
+  void initState() {
+    super.initState();
+    _opacity = new CurvedAnimation(parent: _fade, curve: Curves.ease);
+  }
+
+  void updateState(ScrollableState scrollable) {
+    final ExtentScrollBehavior scrollBehavior = scrollable.scrollBehavior;
```

Comment:
Yes, you're right.  If it's not an ExtentScrollBehavior the update can be short-circuited.